### PR TITLE
add common library use by the maestro project

### PR DESCRIPTION
**Pull request by evonck**
Common library for go.
The godoc links should work as soon as the PR is merged

### CODE REVIEW

File: common/README.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
+# Common
+[![GoDoc](https://godoc.org/github.com/NexwayGroup/nx-lib/common?status.png)](https://godoc.org/github.com/NexwayGroup/nx-lib/common)
+
+The common library provide a generique main class to easily start a new go program.
+It uatomatically set up :
```
Comment (spierre) [inline common/README.md]: automatically

File: common/README.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+   
+GLOBAL OPTIONS:
+   --config, -c "./"		absolute config path [$CONFIG]
+   --help, -h			show help
+   --generate-bash-completion	
+   --version, -v		print the version
+  ```
+  
+  ## Config
+  The genrique ain will also set up a configuration reader. By default the program will try to read from ./config.yml.
```
Comment (spierre) [inline common/README.md]: generic main

File: common/README.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  A --config, -c flag is available so you can define another path to the configuration file.
+  
+   ## Test
+   
+  The library also provide a test package where you can fin a simple test tool to test API endpoints.
```
Comment (spierre) [inline common/README.md]: where you can find

File: common/README.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	test.Init(&common.Handlers{Handle: Handlers})
+	os.Exit(m.Run())
+}
+
+
+func TestingEndpoint(t *testing.T, directory, file string) {
+	w := test.Events(t, "POST", "route", directory, file)
+}
+  ```
+The test.Events function reada file and send its informations to the endpoint of your choice. In order to make sure the test are working you need to define 2 type of file for every test :
```
Comment (spierre) [inline common/README.md]: read a file

File: common/README.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+   help, h	Shows a list of commands or help for one command
+   
+GLOBAL OPTIONS:
+   --config, -c "./"		absolute config path [$CONFIG]
+   --help, -h			show help
+   --generate-bash-completion	
+   --version, -v		print the version
+  ```
+  
+  ## Config
```
Comment (spierre) [inline common/README.md]: remove space in start of line

File: common/README.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  The genrique ain will also set up a configuration reader. By default the program will try to read from ./config.yml.
+  A --config, -c flag is available so you can define another path to the configuration file.
+  
+   ## Test
```
Comment (spierre) [inline common/README.md]: remove space in start of line

File: common/README.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+   
+  The library also provide a test package where you can fin a simple test tool to test API endpoints.
+  
+  ### How To Use
```
Comment (spierre) [inline common/README.md]: remove space in start of line

File: http/decodeEvent.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	"io/ioutil"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// DecodeEventInterface decode a response using interface
+func DecodeEventInterface(resp io.ReadCloser) interface{} {
+	requestData, err := ioutil.ReadAll(resp)
+	var jsonResp interface{}
+	err = json.Unmarshal(requestData, &jsonResp)
```
Comment (UrielCh) [inline http/decodeEvent.go]: use if go syntax.

File: http/decodeEvent.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+		return nil
+	}
+	return jsonResp
+}
+
+// DecodeMapStringInterface decode a response using map[string]interface
+func DecodeMapStringInterface(resp io.ReadCloser) map[string]interface{} {
+	requestData, err := ioutil.ReadAll(resp)
+	jsonResp := make(map[string]interface{})
+	err = json.Unmarshal(requestData, &jsonResp)
```
Comment (UrielCh) [inline http/decodeEvent.go]: Use if go syntax. if x=y; test{}

File: http/decodeEvent.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+	}
+	return jsonResp
+}
+
+// DecodeEventArrayInterface decode a response using []interface
+func DecodeEventArrayInterface(resp io.ReadCloser) []interface{} {
+	requestData, err := ioutil.ReadAll(resp)
+	var jsonResp []interface{}
+	err = json.Unmarshal(requestData, &jsonResp)
+	if err != nil {
```
Comment (UrielCh) [inline http/decodeEvent.go]: Use if go syntax. if x=y; test{}

File: http/apiCaller.go [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+// addQuery add query to the url
+func addQuery(urlValue string, queries map[string]string) string {
+	if queries == nil {
+		return urlValue
+	}
+	urlValue = urlValue + "?"
+	queryNumber := 0
+	for query, value := range queries {
+		urlValue += query + "=" + url.QueryEscape(value)
```
Comment (UrielCh) [inline http/apiCaller.go]: test (queryNumber++ != 0) append &
do not test queryNumber == last.

File: http/apiCaller.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+	payloadJSON, err := json.Marshal(payload)
+	if err != nil {
+		log.Warn("An error occure while decoding the interface: ", err)
+		return nil
+	}
+	req, _ := http.NewRequest(requestType, urlQuery, bytes.NewBuffer(payloadJSON))
+	req = addHeader(req, headers)
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
```
Comment (UrielCh) [inline http/apiCaller.go]: Use if go syntax. if x=y; test{}

File: common/main.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	return lvl
+}
+
+// LoadConfig load configuration file
+func LoadConfig(pathFile, fileName string) {
+	if v == nil {
+		v = viper.New()
+		v.SetConfigName(fileName)
+		v.AddConfigPath(pathFile)
```
Comment (UrielCh) [inline common/main.go]: SetConfigName and AddConfigPath are not compatible
in this case AddConfigPath had no effect

File: common/configTest.yml [hunk 14]
```diff
@@ -0,0 +1,4 @@
+appName: true
+route: steve
+addr: localhost:8081
+log-level: debug
```
Comment (UrielCh) [inline common/configTest.yml]: missing new line

File: common/test/test.go [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+	Router = handlerFunc.Handle()
+	// init serverTest
+	ServerTest = httptest.NewServer(Router)
+	defer ServerTest.Close()
+}
+
+// File2String change the file to a string
+func File2String(filename string) string {
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
```
Comment (UrielCh) [inline common/test/test.go]: Use if go syntax. if x=y; test{}

File: UUID/UUID.go [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+	log "github.com/Sirupsen/logrus"
+)
+
+// UUID Type
+type UUID [16]byte
+
+// NewUUID create a new UUID
+func NewUUID() *UUID {
+	u := &UUID{}
+	_, err := rand.Read(u[:16])
```
Comment (UrielCh) [inline UUID/UUID.go]: use short if syntax

File: common/configTest.yml [hunk 17]
```diff
@@ -0,0 +1,5 @@
...
+
```
Comment (UrielCh) [inline common/configTest.yml]: I said missing New line at end, not an empty line at the end ...

File: common/main.go [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+
+// SetAppVersion set the ap pversion
+func SetAppVersion(version string) {
+	appVersion = version
+}
+
+// SetAppDescription set the app description
+func SetAppDescription(description string) {
+	appDescription = description
+
```
Comment (UrielCh) [inline common/main.go]: an empty line in a Getter ? stange...

File: common/main.go [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// LoadConfig load configuration file
+func LoadConfig(pathFile, fileName string) {
+	if v == nil {
+		v = viper.New()
+		v.SetConfigName(fileName)
+		v.AddConfigPath(pathFile)
+		log.Print("Loading config file")
+		err := v.ReadInConfig() // Find and read the config file
```
Comment (UrielCh) [inline common/main.go]: use compact if syntax

File: common/main.go [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+
+// Run the application
+func Run() {
+	App.Before = func(ctx *cli.Context) error {
+		configPath = ctx.String("config")
+		configInit()
+		return nil
+	}
+	err := App.Run(os.Args)
+	if err != nil {
```
Comment (UrielCh) [inline common/main.go]: use compact if syntax

File: common/test/test.go [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	return string(data)
+}
+
+// APICall call the API
+func APICall(requestType, route, headers, data string) *httptest.ResponseRecorder {
+	request, _ := http.NewRequest(requestType, ServerTest.URL+route, strings.NewReader(data))
+	lines := strings.Split(headers, "\n")
+	for i := range lines {
+		pos := strings.Index(lines[i], ": ")
```
Comment (UrielCh) [inline common/test/test.go]: use the apropriate common string function

File: common/test/test.go [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+func Events(t *testing.T, requestType, route, folder, fileName string) *httptest.ResponseRecorder {
+	files, _ := ioutil.ReadDir(folder)
+	for _, f := range files {
+		if f.Name() == fileName {
+			outmsg := File2String(folder + f.Name())
+			outHeader := File2String(folder + strings.Replace(f.Name(), ".json", ".header", 1))
+			if outmsg == "" {
+				return nil
+			}
+			w := APICall(requestType, route, outHeader, outmsg)
```
Comment (UrielCh) [inline common/test/test.go]: test w value, or remove it

File: http/apiCaller.go [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// addQuery add query to the url
+func addQuery(urlValue string, queries map[string]string) string {
+	if queries == nil {
+		return urlValue
+	}
+	queryNumber := 0
+	for query, value := range queries {
+		if queryNumber == 0 {
```
Comment (UrielCh) [inline http/apiCaller.go]: use queryNumber++ here, not at the end of the loop.

File: common/test/test.go [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+	return string(data)
+}
+
+// APICall call the API
+func APICall(requestType, route, headers map[string]string, data string) *httptest.ResponseRecorder {
+	request, _ := http.NewRequest(requestType, ServerTest.URL+route, strings.NewReader(data))
+	lines := strings.Split(headers, "\n")
+	for i := range lines {
+		headers := strings.SplitN(lines, ":", 2)
+		if strings.Contains(headers[0], "Request") {
```
Comment (UrielCh) [inline common/test/test.go]: merge the 2 if in a single line

File: common/test/test.go [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+	return w
+}
+
+// Events test the event folder
+func Events(t *testing.T, requestType, route, folder, fileName string) *httptest.ResponseRecorder {
+	files, _ := ioutil.ReadDir(folder)
+	for _, f := range files {
+		if f.Name() == fileName {
+			outmsg := File2String(folder + f.Name())
+			outHeader := File2String(folder + strings.Replace(f.Name(), ".json", ".header", 1))
```
Comment (UrielCh) [inline common/test/test.go]: move this line # line down

File: http/apiCaller.go [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+func DELETE(url string, queries map[string]string, headers map[string]string) *http.Response {
+	return callerWithData("DELETE", url, queries, headers, nil)
+}
+
+// addQuery add query to the url
+func addQuery(urlValue string, queries map[string]string) string {
+	queryNumber := 0
+	for query, value := range queries {
+		if queryNumber++ == 0 {
+			urlValue = urlValue + "?"
```
Comment (UrielCh) [inline http/apiCaller.go]: the += is for the dog ?

File: http/apiCaller_test.go [hunk 27]
```diff
@@ -0,0 +1,146 @@
...
+
+func TestAddheader(t *testing.T) {
+	req, _ := http.NewRequest("GET", "http://localhost", nil)
+	header := make(map[string]string)
+	header["Testheader"] = "testValue"
+	req = addHeader(req, header)
+	convey.Convey("req.Header.Testheader should be testValue", t, func() {
+		convey.So(req.Header.Get("Testheader"), convey.ShouldEqual, "testValue")
+	})
+
```
Comment (UrielCh) [inline http/apiCaller_test.go]: oeurfelin empty line

File: http/apiCaller_test.go [hunk 28]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+func Server() {
+	handler := &common.Handlers{Handle: TestingHandlers}
+	router := handler.Handle()
+	ServerTest = httptest.NewServer(router)
+
```
Comment (UrielCh) [inline http/apiCaller_test.go]: empty line

File: http/decodeEvent_test.go [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+)
+
+var (
+	arrayHello []interface{}
+)
+
+func init() {
+	arrayHello = append(arrayHello, hello)
+	arrayHello = append(arrayHello, hello)
+
```
Comment (UrielCh) [inline http/decodeEvent_test.go]: move this empty line 1 line down

File: http/decodeEvent_test.go [hunk 30]
```diff
@@ -0,0 +1,74 @@
...
+	r.POST("/testArray", TestingArrayInterface)
+	r.POST("/testMap", TestingInterface)
+	return r
+}
+
+func Server2() {
+	handler := &common.Handlers{Handle: TestingHandlers2}
+	router := handler.Handle()
+	ServerTest = httptest.NewServer(router)
+
```
Comment (UrielCh) [inline http/decodeEvent_test.go]: bad empty line

File: UUID/UUID_test.go [hunk 31]
```diff
@@ -0,0 +1,19 @@
...
+func TestNewUUID(t *testing.T) {
+	uuid := NewUUID()
+	convey.Convey("uuid should not be nil", t, func() {
+		convey.So(uuid, convey.ShouldNotEqual, nil)
+	})
+	uuid2 := NewUUID()
+	convey.Convey("uuid should not be equl to uuid", t, func() {
+		convey.So(uuid, convey.ShouldNotEqual, uuid2)
+	})
+
```
Comment (UrielCh) [inline UUID/UUID_test.go]: empty line at the end of function => bad

File: http/apiCaller_test.go [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+)
+
+func init() {
+	hello = make(map[string]interface{})
+	hello["test"] = "Hello"
+}
+
+func Testing(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	w.WriteHeader(http.StatusOK)
+	fmt.Fprintln(w, `{ "test" : "Hello"}`)
```
Comment (UrielCh) [inline http/apiCaller_test.go]: remove some spaces

**Pull request merged**